src/store/reader.rs

```diff
@@ -93,6 +93,70 @@ impl StoreReader {
     }
 }
 
+pub struct StoreReaderIterator<'a> {
+    block_offset: usize,
+    document_offset: usize,
+    store_reader: &'a StoreReader,
+    uncompressed_block: Vec<u8>,
+}
+
+impl<'a> StoreReaderIterator<'a> {
+    pub fn has_more_documents(&self) -> bool {
+        self.document_offset < self.uncompressed_block.len()
+    }
+
+    pub fn has_more_blocks(&self) -> bool {
+        self.block_offset < self.store_reader.block_data().len()
+    }
+
+    fn read_block(&mut self) -> crate::Result<()> {
+        let (block_size, compressed_block) = self.store_reader.compressed_block(self.block_offset);
+
+        self.uncompressed_block.clear();
+        decompress(&compressed_block, &mut self.uncompressed_block)?;
+
+        self.block_offset += block_size;
+        self.document_offset = 0;
+        Ok(())
+    }
+}
+
+impl<'a> IntoIterator for &'a StoreReader {
+    type Item = Document;
+    type IntoIter = StoreReaderIterator<'a>;
+
+    /// Iterates over all stored documents.
+    ///
+    /// This method is cheaper than multiple consequent callings of `.get(doc_id)`
+    /// Use it if you need to read all stored documents in the segment
+    fn into_iter(self) -> Self::IntoIter {
+        StoreReaderIterator {
+            block_offset: 0,
+            document_offset: 0,
+            store_reader: &self,
+            uncompressed_block: Vec::new(),
+        }
+    }
+}
+
+impl<'a> Iterator for StoreReaderIterator<'a> {
+    type Item = Document;
+
+    fn next(&mut self) -> Option<Document> {
+        if !self.has_more_documents() {
+            if !self.has_more_blocks() {
+                return None;
+            }
+            self.read_block().unwrap();
+        }
+        let mut cursor = io::Cursor::new(&self.uncompressed_block[self.document_offset..]);
```

Comment:
I was a rust rookie when I started tantivy. I suspect you copied this code from somewhere else.
io::Cursor is not useful here since `&[u8]` implements read.